File: packages/image-labeling/android/src/main/java/io/capawesome/capacitorjs/plugins/mlkit/imagelabeling/ImageLabelingPlugin.java
package io.capawesome.capacitorjs.plugins.mlkit.imagelabeling;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import io.capawesome.capacitorjs.plugins.mlkit.imagelabeling.classes.options.ProcessImageOptions;
import io.capawesome.capacitorjs.plugins.mlkit.imagelabeling.classes.results.ProcessImageResult;
import io.capawesome.capacitorjs.plugins.mlkit.imagelabeling.interfaces.NonEmptyResultCallback;
import io.capawesome.capacitorjs.plugins.mlkit.imagelabeling.interfaces.Result;

@CapacitorPlugin(name = "ImageLabeling")
public class ImageLabelingPlugin extends Plugin {

    public static final String TAG = "ImageLabeling";
    public static final String ERROR_PROCESS_IMAGE_CANCELED = "processImage canceled.";
    public static final String ERROR_PATH_MISSING = "path must be provided.";
    public static final String ERROR_LOAD_IMAGE_FAILED = "image could not be loaded.";
    public static final String ERROR_UNKNOWN_ERROR = "An unknown error occurred.";

    private ImageLabeling implementation;

    @Override
    public void load() {
        implementation = new ImageLabeling(this);
    }

    @PluginMethod
    public void processImage(PluginCall call) {
        try {
            ProcessImageOptions options = new ProcessImageOptions(call);
            NonEmptyResultCallback<ProcessImageResult> callback = new NonEmptyResultCallback<>() {
                @Override
                public void success(@NonNull ProcessImageResult result) {
                    resolveCall(call, result);
                }

                @Override
                public void error(@NonNull Exception exception) {
                    rejectCall(call, exception);
                }
            };

            implementation.processImage(options, callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    private void rejectCall(@NonNull PluginCall call, @NonNull Exception exception) {
        String message = exception.getMessage();
        if (message == null) {
            message = ERROR_UNKNOWN_ERROR;
        }
        Logger.error(TAG, message, exception);
        call.reject(message);
    }

    private void resolveCall(@NonNull PluginCall call, @Nullable Result result) {
        if (result == null) {
            call.resolve();
        } else {
            call.resolve(result.toJSObject());
        }
    }
}

File: packages/image-labeling/ios/Plugin/ImageLabelingPlugin.swift
import Foundation
import Capacitor
import MLKitVision
import MLKitImageLabeling

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(ImageLabelingPlugin)
public class ImageLabelingPlugin: CAPPlugin {
    public let tag = "ImageLabeling"

    private var implementation: ImageLabeling?

    override public func load() {
        implementation = ImageLabeling(plugin: self)
    }

    @objc func processImage(_ call: CAPPluginCall) {
        do {
            let options = try ProcessImageOptions(call)

            implementation?.processImage(options, completion: { result, error in
                if let error = error {
                    self.rejectCall(call, error)
                } else {
                    self.resolveCall(call, result)
                }
            })
        } catch {
            self.rejectCall(call, error)
        }
    }

    private func rejectCall(_ call: CAPPluginCall, _ error: Error) {
        CAPLog.print("[", self.tag, "] ", error)
        call.reject(error.localizedDescription)
    }

    private func resolveCall(_ call: CAPPluginCall, _ result: Result?) {
        if let result = result?.toJSObject() as? JSObject {
            call.resolve(result)
        } else {
            call.resolve()
        }
    }
}
